File: 2022/day_02/haskell/app/Game.hs
module Game where

type Points = Int

data Shape = Paper | Rock | Scissors
data Outcome = Lose | Draw | Win
data Round = RoundPart1 (Shape, Shape) | RoundPart2 (Shape, Outcome)

-- Parses a given input file line and converts it into a round.
lineToRound :: Int -> String -> Round
lineToRound 1 line = RoundPart1 (charToShape $ head line, charToShape $ line !! 2)
lineToRound 2 line = RoundPart2 (charToShape $ head line, charToOutcome $ line !! 2)

-- Converts an input character into a shape.
charToShape :: Char -> Shape
charToShape 'A' = Rock
charToShape 'X' = Rock
charToShape 'B' = Paper
charToShape 'Y' = Paper
charToShape 'C' = Scissors
charToShape 'Z' = Scissors

-- Converts an input character into an outcome.
charToOutcome :: Char -> Outcome
charToOutcome 'X' = Lose
charToOutcome 'Y' = Draw
charToOutcome 'Z' = Win

-- Converts a round into an outcome for part 1 of the problem.
roundToOutcome :: Round -> Outcome
roundToOutcome (RoundPart1 (Rock, Paper)) = Win
roundToOutcome (RoundPart1 (Rock, Scissors)) = Lose
roundToOutcome (RoundPart1 (Paper, Rock)) = Lose
roundToOutcome (RoundPart1 (Paper, Scissors)) = Win
roundToOutcome (RoundPart1 (Scissors, Rock)) = Win
roundToOutcome (RoundPart1 (Scissors, Paper)) = Lose
roundToOutcome (RoundPart1 (_, _)) = Draw

-- Converts a round into a shape for part 2 of the problem.
roundToShape :: Round -> Shape
roundToShape (RoundPart2 (Rock, Lose)) = Scissors
roundToShape (RoundPart2 (Rock, Win)) = Paper
roundToShape (RoundPart2 (Paper, Lose)) = Rock
roundToShape (RoundPart2 (Paper, Win)) = Scissors
roundToShape (RoundPart2 (Scissors, Lose)) = Paper
roundToShape (RoundPart2 (Scissors, Win)) = Rock
roundToShape (RoundPart2 (s, Draw)) = s

-- Calculates the points won from a given outcome.
outcomeToPoints :: Outcome -> Points
outcomeToPoints Lose = 0
outcomeToPoints Draw = 3
outcomeToPoints Win = 6

-- Calculates the points won from playing a given shape.
playToPoints :: Shape -> Points
playToPoints Rock = 1
playToPoints Paper = 2
playToPoints Scissors = 3

-- Calculates the points won from a given round.
getRoundPoints :: Round -> Points
getRoundPoints (RoundPart1 round) = outcomeToPoints (roundToOutcome (RoundPart1 round)) + playToPoints (snd round)
getRoundPoints (RoundPart2 round) = outcomeToPoints (snd round) + playToPoints (roundToShape (RoundPart2 round))

File: 2022/day_02/rust/src/game.rs
type Points = u32;

pub enum Shape {
    Paper,
    Rock,
    Scissors,
}
pub enum Outcome {
    Lose,
    Draw,
    Win,
}
pub enum Round {
    RoundPart1(Shape, Shape),
    RoundPart2(Shape, Outcome),
}

/// Parses a given vector of input file lines and converts them into a vector of rounds.
pub fn lines_to_rounds(lines: &Vec<String>, problem_part: u8) -> Vec<Round> {
    let mut rounds: Vec<Round> = Vec::new();
    for line in lines {
        let first = char_to_shape(&line[..1]);
        match problem_part {
            1 => {
                let second = char_to_shape(&line[2..]);
                let round = Round::RoundPart1(first, second);
                rounds.push(round);
            }
            _ => {
                let second = char_to_outcome(&line[2..]);
                let round = Round::RoundPart2(first, second);
                rounds.push(round);
            }
        }
    }
    rounds
}

/// Converts an input character into a shape.
fn char_to_shape(character: &str) -> Shape {
    match character {
        "A" => Shape::Rock,
        "X" => Shape::Rock,
        "B" => Shape::Paper,
        "Y" => Shape::Paper,
        &_ => Shape::Scissors,
    }
}

/// Converts an input character into an outcome.
fn char_to_outcome(character: &str) -> Outcome {
    match character {
        "X" => Outcome::Lose,
        "Y" => Outcome::Draw,
        &_ => Outcome::Win,
    }
}

/// Converts a round into an outcome for part 1 of the problem.
fn round_to_outcome(round: &Round) -> Outcome {
    match round {
        Round::RoundPart1(Shape::Rock, Shape::Paper) => Outcome::Win,
        Round::RoundPart1(Shape::Rock, Shape::Scissors) => Outcome::Lose,
        Round::RoundPart1(Shape::Paper, Shape::Scissors) => Outcome::Win,
        Round::RoundPart1(Shape::Paper, Shape::Rock) => Outcome::Lose,
        Round::RoundPart1(Shape::Scissors, Shape::Rock) => Outcome::Win,
        Round::RoundPart1(Shape::Scissors, Shape::Paper) => Outcome::Lose,
        _ => Outcome::Draw,
    }
}

/// Converts a round into a shape for part 2 of the problem.
fn round_to_shape(round: Round) -> Shape {
    match round {
        Round::RoundPart2(Shape::Rock, Outcome::Lose) => Shape::Scissors,
        Round::RoundPart2(Shape::Rock, Outcome::Win) => Shape::Paper,
        Round::RoundPart2(Shape::Paper, Outcome::Win) => Shape::Scissors,
        Round::RoundPart2(Shape::Scissors, Outcome::Lose) => Shape::Paper,
        Round::RoundPart2(s, Outcome::Draw) => s,
        _ => Shape::Rock,
    }
}

/// Calculates the points won from a given outcome.
pub fn outcome_to_points(outcome: &Outcome) -> Points {
    match outcome {
        Outcome::Lose => 0,
        Outcome::Draw => 3,
        Outcome::Win => 6,
    }
}

/// Calculates the points won from playing a given shape.
fn play_to_points(shape: &Shape) -> Points {
    match shape {
        Shape::Rock => 1,
        Shape::Paper => 2,
        Shape::Scissors => 3,
    }
}

/// Calculates the points won from a given round.
pub fn get_round_points(round: Round) -> Points {
    match round {
        Round::RoundPart1(_, ref y) => {
            outcome_to_points(&round_to_outcome(&round)) + play_to_points(y)
        }
        Round::RoundPart2(_, ref y) => {
            outcome_to_points(&y) + play_to_points(&round_to_shape(round))
        }
    }
}

/// Sums the total points won from a given vector of rounds.
pub fn sum_points_for_rounds(rounds: Vec<Round>) -> Points {
    let mut total_points: Points = 0;
    for round in rounds {
        total_points += get_round_points(round);
    }
    total_points
}
